python/python/Ice/ObjectPrx.py

```diff
@@ -40,16 +58,31 @@ def uncheckedCast(type, proxy, facet=None):
     return IcePy.ObjectPrx.newProxy(type, proxy)
 
 
-def checkedCast(type, proxy, facet=None, context: dict[str, str] | None = None):
+@overload
+def checkedCast(
+    type: Type[T], proxy: ObjectPrx, facet: str | None = None, context: dict[str, str] | None = None
+) -> T: ...
+
+
+@overload
+def checkedCast(
+    type: Type[T], proxy: None, facet: str | None = None, context: dict[str, str] | None = None
+) -> None: ...
+
+
+def checkedCast(
+    type: Type[T], proxy: ObjectPrx | None, facet: str | None = None, context: dict[str, str] | None = None
+) -> T | None:
     """
     Downcasts a proxy after confirming the target object's type via a remote invocation.
 
     Parameters
     ----------
     type : type
         The proxy target type.
-    proxy : ObjectPrx
-        The source proxy (can be None).
+
+    proxy : ObjectPrx or None
```

Comment:
I think we should switch to `ObjectPrx | None` for handwritten code, generated code would use the same for type hint and docs. For a follow up PR.